### PR TITLE
[Feat] Add multi-reward advantage computation (GDPO)

### PR DESCRIPTION
**Pull request by shyampathak22**
Implements per-reward normalization as per [Group reward-Decoupled Normalization Policy Optimization for Multi-reward RL Optimization](https://arxiv.org/html/2601.05242v1#S4.F6), enabling RL training with multiple reward signals without reward collapse. When using standard GRPO with multiple rewards, different reward combinations can collapse to identical advantage values. GDPO normalizes each reward signal independently within its problem group before aggregation, preserving relative signal of each reward type.

Verifiers already returns a rollout["metrics"] value, from which the computation pulls individual metrics, normalizes each separately, then aggregates. This pathway is only active when `reward_keys=[]`  is set in the environment config, as follows:

```
[[orchestrator.env]]
id = "shyampathak/deepscaler-multireward"
name = "deepscaler_multireward"
reward_keys = ["correctness_reward", "length_reward"] # This activates the multi-reward computation path
reward_weights = [1.0, 0.5]  # Optional, defaults to equal weights

[orchestrator.advantage]
batch_normalize = true  # Apply batch-wise normalization
std_eps = 1e-8 # Batch norm std 
```

To test, multi-reward environments for both [DeepScaleR](https://app.primeintellect.ai/dashboard/environments/shyampathak/deepscaler-multireward) and [GSM8K](https://app.primeintellect.ai/dashboard/environments/shyampathak/gsm8k-multireward) 
were created. All below experiments were done using the GSM8K environment, due to time and (mostly) compute constraints for the large scale DeepScaleR runs.

---

> [!NOTE]
> Enables multi-reward RL with configurable normalization and adaptive weighting, and wires it through orchestrator, config, logging, and checkpoints.
> 
> - Adds `compute_advantages_multi_reward` supporting GDPO (`per_reward_normalize=True`) and GRPO (`False`) with optional batch-wise normalization and `std_eps`
> - Introduces `EnvConfig.reward_keys`/`reward_weights` and `AdvantageConfig` fields (`batch_normalize`, `per_reward_normalize`, `std_eps`); new `AdaptiveWeightConfig`
> - Implements `AdaptiveWeightManager` to EMA-track reward saturation and decay weights with ratchet/recovery; exposes state for checkpointing
> - Orchestrator: initializes per-env adaptive managers, computes per-env advantages for mixed single/multi-reward setups, logs `adaptive_weights/*` and `reward_ema/*`, and saves/restores adaptive weight state via `Progress.adaptive_weight_states`
> - Updates `ckpt.Progress` to include adaptive weight state; adds unit tests for advantage and adaptive weights; updates `.gitignore` and `CHANGELOG.md`
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c0b629013d03107054a34e1f56150b0671ba56ca. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by shyampathak22** [comment]
Added an Adaptive Reward Weighting mechanism to reduce the sparsity axis presented by reward gating.

While reward gating is good for the purposes of preventing reward hacking, ARW + GDPO shows that its fine to let the model learn a trivial reward early on. To mitigate this, I added a decay to the weight of a given reward, which dynamically adjusts as the running mean of a given reward approaches saturation. A hysteresis mechanism and slow-leak ratcheting mechanism are in place to prevent oscillation between optimization of rewards. 

 Interestingly, in my testing, GRPO + ARW seems to not have nearly as good of an effect. The decoupled nature of GDPO's reward seems very necessary for improvements. Results comparison a GRPO + Gated Rewards with a GDPO + Adaptive Reward Weighting as well as a GRPO with Adaptive Reward Weighting are shown below. 

Screenshot 2026-01-24 005432

Training Rewards for Correctness and Length (Must be below 384 token output for length reward). All 3 models immediately are capable of maximizing length reward, However, the gated GRPO gets a much sparser signal. Furthermore, both GRPO runs suffer from a confounded reward signal, where the GDPO doesn't, which can be seen in its more stable curves. Both the Length reward and the Correctness have less variation throughout training.

Screenshot 2026-01-24 005407

Screenshot 2026-01-24 005414

Above shows the evaluation results of both GRPO + ARW and GDPO + ARW. We can see that the accuracy was nearly the same, but GDPO + ARW achieved a 0.60995, while GRPO + ARW achieved a 0.57316 score using the primeintellect/math-env evaluation with the test set of openai/gsm8k. Both were done with greedy decoding and pass@1 metric reported.

The interesting part is that GRPO + ARW seemed to forget towards the end of its training how to maintain its length reward. In doing so, it began to, while maintaining its accuracy, lose a grip on its length capabilities. GDPO + ARW averaged 201 tokens per completion, while GRPO + ARW averaged 1140. The difference is stark and clear! GDPO + Adaptive reward weighting allows for cleaner and clearer signals of rewards, while allowing the density presented by multiple rewards to remain while continuing improvement via a shifted focus through the weighting mechanisms!
(PS, I was unable to eval on GRPO-gated model, but given the nearly 8 point gain from GDPO + ARW over it, I didn't prioritize this, but it is running right now)

also i apologize for like 500 lines of test code, i was paranoid about breaking GRPO in the process lol
